src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -343,7 +347,77 @@ public void Stop()
             }
         }
 
+
+        /* W3C support functionality (see https://w3c.github.io/trace-context) */
+
+        /// <summary>
+        /// Holds the W3C 'tracestate' header.   This is typically used by logging systems
+        /// to store vendor-specific information that must flow.
+        /// Logically it is just a kind of baggage (if flows just like baggage), but because
+        /// it is expected to be special cased (it has its own HTTP header), it is more 
+        /// convenient/efficient if it is not lumped in with other baggage.   
+        /// </summary>
+        public string TraceState
+        {
+            get
+            {
+                for (var activity = this; activity != null; activity = activity.Parent)
+                {
+                    var val = activity._traceState;
+                    if (val != null)
+                        return val;
+                }
+                return null;
+            }
+            set
+            {
+                _traceState = value;
+            }
+        }
+  
+        /// <summary>
+        /// Activity tries to use the same format for IDs as its parent it has a parent.  
+        /// However if the activity has no parent, it has to do something.   
+        /// This sets it so that WC3 format is used in this case.  
+        /// </summary>
+        static public bool UseW3CFormat { get; set; }
+
+        /// <summary>
+        /// Returns true if the ID happens to be in W3C format XX-XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX-XXXXXXXXXXXXXXXX-XX
+        /// this is a convenience method.  
+        /// </summary>
+        public bool IsW3CFormat { get; private set; }
+
         #region private 
+        /// <summary>
+        /// Returns true if 'id' has the format of a WC3 id see https://w3c.github.io/trace-context
+        /// </summary>
+        private static bool IsW3CId(string id)
+        {
+            // A W3CId is  
+            //  * 2 chars Version
+            //  * 32 chars traceId
+            //  * 16 chars spanId
+            //  * 2 chars flags 
+            //  * 3 chars - separators.  
+            //  = 55 chars (see https://w3c.github.io/trace-context)
+            // We require that all non-WC3IDs NOT start with a digit.  
+            return id.Length == 55 && char.IdDigit(id[0]);
+        }
+        private string GenerateW3CId()
+        {
+            string newSpanId = Guid.NewGuid().ToString("n").Substring(16, 16);
+            if (ParentId != null)
+            {
+                return ParentId.SubString(0, 36) + newSpanId + ParentId.SubString(52, 3);
```

Comment:
I realize this is a strawman, but just want to clarify that **we want to have access to newSpanId and parentSpanId without parsing a string**. As we want to decouple http wire format from how telemetry is reported. And we'll potentially access it many times during Activity lifetime, so it should be more efficient to keep them rather than calculate each time.

Reply:
Assuming there was Request-Id in the incoming request and SetParentId was called with |abc...1.2.3, rootId may become invalid

We should validate ParentId here and if it's not valid one, we should generate a new traceId

Reply:
I am interested in insure we don't have a bunch of unnecessary string maniputation, so yes, I am open to having activity cache/parse things, but I want to understand the issue a bit more first.  In particualr I want to understand why you want SpanIds AT RUNTIME.   I can see wanting them at PROCESSING TIME, but otherwise it is not clear to me (seems like you pass the ID around.   

It actually seems MORE VERSION RESILIANT, if the RUNTIME part of the system treats to the extent possible as a opaque blob (which is what Activity.Id is),   This allows the format of the ID to change (add more information) without breaking all existing code.   Thus ideally you should not be parsing the ID, and certainly not multiple times.    Can you describe when you think you need to parse the spans?

Reply:
> We should validate ParentId here and if it's not valid one, we should generate a new traceId

The logic behind the code is that in SetParentId, if we detect that the ID is a valid W3CId we set the 'IsW3CFormat' bit.  We only come into this code in the W3CId case.

In your scneario (where the incomming request is using the old format, we don't ever run this code (we use the old format and do the old thing).     

The expectation is that systems that wish to only support the new W3C format will test the IsW3CFormat bit and drop the activity if it is not true.

Reply:
>   I want to understand why you want SpanIds AT RUNTIME. I can see wanting them at PROCESSING TIME, but otherwise it is not clear to me (seems like you pass the ID around.

> It actually seems MORE VERSION RESILIANT, if the RUNTIME part of the system treats to the extent possible as a opaque blob (which is what Activity.Id is), This allows the format of the ID to change (add more information) without breaking all existing code. Thus ideally you should not be parsing the ID, and certainly not multiple times. Can you describe when you think you need to parse the spans?


When tracing system reports telemetry (it's applicable to AppInsights, OpenCensus, Zipkin, Jaeger, NewRelic and Dynatrace and probably the rest of them) it records traceId, parentSpanId and spanId separately. So at least at the moment of recoding telemetry item we should have them all. In reality, we have more then one telemetry reported in scope of the activity: log, exception or some event that requires traceId and spanId to be properly set on such telemetry as well for correlation.   

This ids are used for indexing and sql queries on the backends. 

Does it explain my point?

Reply:
and yes, we could actually save whole W3CId as span id, but this is redundunt as we save  traceId in the other field anyway.

Reply:
> The logic behind the code is that in SetParentId, if we detect that the ID is a valid W3CId we set the 'IsW3CFormat' bit. We only come into this code in the W3CId case.

We generate W3C Id if DefaultIdFormat. Which we are going to set tot true in all possible cases.
```csharp
                if (IsW3CId || DefaultIdFormat == ActivityIDFormat.W3C)
                {
                    IsW3CFormat = true;
                    Id = GenerateW3CId();
                }
```

Reply:
You found a bug.  The logic should be 
```
                if (IsW3CId || ParentId == null && DefaultIdFormat == ActivityIDFormat.W3C)
```
I have updated the PR.

Reply:
This is a feature. With && it creates even more complications because now the upstream service controls the protocol and tracing system or user cannot prevent it from happening.

let's assume service X runs on AspNetCore 2.0 with W3C Activity. 
It is called by A and B upstream services and it calls C and D  downstream services.

Let's assume X upgrades to corefx 3.0, but others are 2.0. Things are good, everything works per over old protocol.

Service X does not necessarily control how A, B, C, or D update. It's enough for any of the upstream services to update to break correlation from A to C or D: 

After that A updates to 3.0 and then requests coming from  A to X are traceable, but not further.
Request from B are traceable end-to-end, 

As we don't know customer's layout we have to provide them with a way to force W3C or force Request-Id.  Following upstream protocol does not help anyone,

My proposal is to 
- allow to force W3C ids generation in all cases (even if parent is request-id)
- emit W3C header in all cases (always or when forced) 
- emit Request-Id in format `|traceid.spanid.` (always or when enabled, or leave it to the tracing system)
- generate old ids only when request-id is enforced, change RootId algo generation to be the same as traceId, so that reasonable mapping from Activity.Id to w3c would be possible anyway

X decides to update. It forces W3C and emits Request-Id with `|traceid.spanid` value. Correlation with all downstream services works.
This does not require downstream customers to even update AspNetCore/HttpClient.

So all customers start to upgrade from the topmost service/client they own, and correlation works and dependency calls troubleshooting works in all cases. This way they control their part of infra and not controlled by parts they don't own.

Correlation with upstream services will be broken unless tracing system can support multiple parents for the same request - which we just enabled.

However depending on the layout, customers may choose to turn W3C off and update bottom-up.

Reply:
> When tracing system reports telemetry (it's applicable to AppInsights, OpenCensus, Zipkin, Jaeger, NewRelic and Dynatrace and probably the rest of them) it records traceId, parentSpanId and spanId separately

I understand.   What are the data types for this data?   Presumably spanIds are passed as a 64 bit long?   But the TraceId is too big for that.  What do they use in their signature to represent it?

Reply:
AI - strings
OpenCensus - byte[]
Zipkin (and perhaps Jaeger) - strings.

I can check for the rest, but let's assume it's strings

Reply:
Thanks for the info on how TraceId is encoded.   I am assuming at this point that logging systems don't care about the Flags part of the ID (assuming we have 'Recorded' bit on Activity).   

I need to think about your feedback on the disadvantages of following the convention of the incoming request.

Reply:
One other quick question.   The TraceId is a string.  Is the string just a 128bit number in Hex?   (I am assuming yes).

Reply:
it's 16 bytes, i.e. 128 bit number

Reply:
Right,  a 128 bit (16 byte) number would take 32 characters to express as a string if you use Hex (2 chars per byte).   Is that what they do?

Reply:
Exactly, thats what they do